lib/queue.go

```diff
@@ -0,0 +1,44 @@
+package lib
+
+import (
+	"fmt"
+)
+
+var (
+	// ErrNotFound is a standard error for a missing thing
+	ErrNotFound = fmt.Errorf("not found")
+)
+
+// Queue is an interface for queing up Requests. it's expected that the queue
+// operates in FIFO order, pushing requests that need processing onto one end
+// popping requests off the other for processing
+type Queue interface {
+	Push(*Request)
+	Pop() *Request
+	Len() (int, error)
+	Chan() (chan *Request, error)
+}
+
+// MemQueue is an in-memory implementation of the Queue interface,
+// it's just a fancy channel
+type MemQueue chan *Request
+
+// Push adds a fetch request to the end of the queue
+func (q MemQueue) Push(t *Request) {
+	q <- t
+}
+
+// Pop removes a request from the queue
+func (q MemQueue) Pop() *Request {
+	return <-q
```

Comment:
May possibly want to implement acknowledgement/confirmation for guaranteed delivery here (when popping, move the item to a secondary queue, then delete it from that queue when acknowledgement happens or move it back to the main queue if you don’t get acknowledgement within a given timeframe because the worker died).

No need to do this now, but probably add a TODO or some note about it.

Reply:
Good point. I could see a potential bug arising from the fact that we don't update the RequestStore when popping off the queue to reflect that a URL is being worked on because we don't have explicit acknowledgement, so things may get stuck in a "in queue" state when they have in fact failed.

Either we need explicit ACK, or we'd have to trust workers to _always_ produce a resource-as-response.

Reply:
Probably an explicit ACK — since the queue is in charge of receiving the ACK and it knows when the message was popped, it can make a good judgement about whether the work has taken too long and potentially died. The coordinator only knows when something was put *on* the queue, so it can’t make a very good judgement about when/whether to give up and re-add it.